src/datasets/arrow_writer.py

```diff
@@ -432,8 +432,17 @@ def write_examples_on_file(self):
         )
         batch_examples = {}
         for col in cols:
-            # Since current_examples contains (example, key) tuples
-            batch_examples[col] = [row[0][col] for row in self.current_examples]
+            # We use row[0][col] since current_examples contains (example, key) tuples.
+            # Morever, examples could be Arrow arrays of 1 element.
+            # This can happen in `.map()` when we want to re-write the same Arrow data
+            if all(isinstance(row[0][col], (pa.Array, pa.ChunkedArray)) for row in self.current_examples):
+                arrays = [row[0][col] for row in self.current_examples]
+                batch_examples[col] = array_concat(arrays)
+            else:
+                batch_examples[col] = [
+                    row[0][col].to_pylist()[0] if isinstance(row[0][col], (pa.Array, pa.ChunkedArray)) else row[0][col]
+                    for row in self.current_examples
+                ]
```

Comment:
Nit: We could avoid the exhaustive `isinstance` check with `try/except`, but I don't think the peformance gain significant enough: 

```suggestion
            if isinstance(self.current_examples[0][0][col], (pa.Array, pa.ChunkedArray)):
                try:
                    batch_examples[col] = array_concat([row[0][col] for row in self.current_examples])
                except Exception:
                    batch_examples[col] = [
                        row[0][col].to_pylist()[0] if isinstance(row[0][col], (pa.Array, pa.ChunkedArray)) else row[0][col]
                        for row in self.current_examples
                ]
                    
            else:
                batch_examples[col] = [
                    row[0][col].to_pylist()[0] if isinstance(row[0][col], (pa.Array, pa.ChunkedArray)) else row[0][col]
                    for row in self.current_examples
                ]
```

Reply:
ok !

Reply:
`except Exception` would make debugging harder in the scenarios when a relevant error is raised inside `array_concat`, so I think we can leave this code as-is